File: python_modules/jupedsim/tests/test_agent_view.py
# SPDX-License-Identifier: LGPL-3.0-or-later
from dataclasses import dataclass, replace

import jupedsim as jps
import pytest
import shapely

_STAY_PUT = (0.0, 0.0)


@dataclass(kw_only=True, frozen=True)
class _State:
    group: int = 0
    probe: bool = False


def _open_room():
    return shapely.Polygon([(0, 0), (20, 0), (20, 20), (0, 20)])


def _make_sim(model, geometry=None):
    if geometry is None:
        geometry = _open_room()
    sim = jps.Simulation(model=model, geometry=geometry, dt=0.05)
    exit_id = sim.add_exit_stage([(19, 9), (19, 11), (20, 11), (20, 9)])
    journey_id = sim.add_journey(jps.JourneyDescription([exit_id]))
    return sim, exit_id, journey_id


def _add_agent(sim, journey_id, stage_id, position, group=0, probe=False):
    return sim.add_agent(
        journey_id=journey_id,
        stage_id=stage_id,
        position=position,
        state=_State(group=group, probe=probe),
    )


def _walled_room():
    """20x20 room with a thin vertical wall from y=0 to y=15 at x=10.
    Agents on the left (x<9.9) cannot see across to the right (x>10.1)
    below y=15; above y=15 the gap gives free sight lines."""
    outer = shapely.Polygon([(0, 0), (20, 0), (20, 20), (0, 20)])
    wall = shapely.Polygon([(9.9, 0), (10.1, 0), (10.1, 15), (9.9, 15)])
    return outer.difference(wall)


class _CapturingModel(jps.CustomOperationalModel):
    """Calls other_agents_in_range for the probe agent and records results."""

    def __init__(self, *, radius=5.0, predicate=None):
        """predicate is called as ``(step, neighbor) -> bool``."""
        super().__init__()
        self._radius = radius
        self._predicate = predicate
        self.neighbors: list[tuple] = []
        self.predicate_neighbors: list[tuple] = []

    def compute_next_state(self, state, step):
        if state.probe:
            self.predicate_neighbors.clear()
            boundaries = step.walls_nearby()

            def _tracking(neighbor):
                self.predicate_neighbors.append(neighbor.relative_position)
                if self._predicate is None:
                    return True
                return self._predicate(step, neighbor, boundaries)

            neighbors = step.other_agents_in_range(self._radius, _tracking)
            self.neighbors = [n.relative_position for n in neighbors]
        return replace(state), _STAY_PUT


# ---------------------------------------------------------------------------
# Basic filter tests
# ---------------------------------------------------------------------------


def test_no_predicate_returns_all_in_radius():
    model = _CapturingModel(radius=5.0)
    sim, exit_id, journey_id = _make_sim(model)
    _add_agent(sim, journey_id, exit_id, (2.0, 10.0), probe=True)
    _add_agent(sim, journey_id, exit_id, (3.0, 10.0))  # 1.0 m away — in range
    _add_agent(sim, journey_id, exit_id, (4.0, 10.0))  # 2.0 m away — in range
    _add_agent(
        sim, journey_id, exit_id, (15.0, 10.0)
    )  # 13.0 m away — out of range

    sim.iterate()

    assert len(model.neighbors) == 2
    assert (1.0, 0.0) in model.neighbors
    assert (2.0, 0.0) in model.neighbors


def test_reject_all_predicate_returns_empty():
    model = _CapturingModel(
        radius=5.0, predicate=lambda step, n, boundaries: False
    )
    sim, exit_id, journey_id = _make_sim(model)
    _add_agent(sim, journey_id, exit_id, (2.0, 10.0), probe=True)
    _add_agent(sim, journey_id, exit_id, (3.0, 10.0))
    _add_agent(sim, journey_id, exit_id, (4.0, 10.0))

    sim.iterate()

    assert model.neighbors == []


def test_group_filter_returns_only_matching_group():
    model = _CapturingModel(
        radius=5.0, predicate=lambda step, n, boundaries: n.state.group == 1
    )
    sim, exit_id, journey_id = _make_sim(model)
    _add_agent(
        sim, journey_id, exit_id, (2.0, 10.0), group=1, probe=True
    )  # probe — not a neighbor of itself
    _add_agent(
        sim, journey_id, exit_id, (3.0, 10.0), group=1
    )  # same group → kept
    _add_agent(
        sim, journey_id, exit_id, (4.0, 10.0), group=0
    )  # different group → filtered out
    _add_agent(
        sim, journey_id, exit_id, (2.0, 11.0), group=1
    )  # same group → kept

    sim.iterate()

    assert len(model.neighbors) == 2
    assert (1.0, 0.0) in model.neighbors
    assert (0.0, 1.0) in model.neighbors


def test_predicate_never_called_with_self():
    # accept-all predicate — only here to trigger tracking
    model = _CapturingModel(
        radius=100.0, predicate=lambda step, n, _walls: True
    )
    sim, exit_id, journey_id = _make_sim(model)
    _add_agent(sim, journey_id, exit_id, (2.0, 10.0), probe=True)
    _add_agent(sim, journey_id, exit_id, (3.0, 10.0))

    sim.iterate()

    assert (0.0, 0.0) not in model.predicate_neighbors


# ---------------------------------------------------------------------------
# no_geometry_between tests
# ---------------------------------------------------------------------------


def _visible(step, neighbor, boundaries):
    return step.no_geometry_between(neighbor.relative_position, boundaries)


def test_no_geometry_between_filters_occluded_agents():
    """Agents behind the wall should be excluded by no_geometry_between."""
    model = _CapturingModel(radius=20.0, predicate=_visible)
    sim, exit_id, journey_id = _make_sim(model, geometry=_walled_room())
    _add_agent(sim, journey_id, exit_id, (5.0, 10.0), probe=True)
    _add_agent(sim, journey_id, exit_id, (5.0, 5.0))  # same side, visible
    _add_agent(sim, journey_id, exit_id, (15.0, 10.0))  # behind wall, occluded

    sim.iterate()

    assert (0.0, -5.0) in model.neighbors
    assert (10.0, 0.0) not in model.neighbors


def test_no_geometry_between_agent_above_wall_is_seen():
    """Agents whose connecting line stays above the wall top (y>15) are visible.

    Probe is at (5, 18) and target at (15, 18): the horizontal path y=18 never
    crosses the wall (which ends at y=15), so the target must be returned.
    """
    model = _CapturingModel(radius=20.0, predicate=_visible)
    sim, exit_id, journey_id = _make_sim(model, geometry=_walled_room())
    _add_agent(
        sim, journey_id, exit_id, (5.0, 18.0), probe=True
    )  # probe — above wall top
    _add_agent(
        sim, journey_id, exit_id, (15.0, 18.0)
    )  # far side, same height → visible

    sim.iterate()

    assert (10.0, 0.0) in model.neighbors


def test_composed_no_geometry_between_and_group_filter():
    """Combining no_geometry_between with a group predicate using a lambda."""
    model = _CapturingModel(
        radius=20.0,
        predicate=lambda step, n, boundaries: (
            _visible(step, n, boundaries) and n.state.group == 1
        ),
    )
    sim, exit_id, journey_id = _make_sim(model, geometry=_walled_room())
    _add_agent(
        sim, journey_id, exit_id, (5.0, 10.0), group=0, probe=True
    )  # probe
    _add_agent(
        sim, journey_id, exit_id, (5.0, 15.0), group=1
    )  # visible, group 1 → kept
    _add_agent(
        sim, journey_id, exit_id, (5.0, 5.0), group=0
    )  # visible, group 0 → filtered
    _add_agent(
        sim, journey_id, exit_id, (15.0, 10.0), group=1
    )  # occluded, group 1 → filtered

    sim.iterate()

    assert model.neighbors == [(0.0, 5.0)]


# ---------------------------------------------------------------------------
# Wall views
# ---------------------------------------------------------------------------


class _WallCapturingModel(jps.CustomOperationalModel):
    """Records the walls the probe agent sees."""

    def __init__(self, *, distance=3.0):
        super().__init__()
        self._distance = distance
        self.walls: list = []

    def compute_next_state(self, state, step):
        if state.probe:
            self.walls = step.walls_in_range(self._distance)
        return replace(state), _STAY_PUT


def test_wall_view_carries_the_projection_onto_the_wall():
    model = _WallCapturingModel(distance=3.0)
    sim, exit_id, journey_id = _make_sim(model, geometry=_walled_room())
    _add_agent(sim, journey_id, exit_id, (8.0, 5.0), probe=True)

    sim.iterate()

    assert model.walls
    nearest = min(model.walls, key=lambda w: w.distance)

    # The near face of the wall sits at x = 9.9, so 1.9 to the agent's right.
    assert nearest.distance == pytest.approx(1.9)
    assert nearest.closest_point == pytest.approx((1.9, 0.0))
    # The normal points from the wall back at the agent, i.e. against +x here.
    assert nearest.normal == pytest.approx((-1.0, 0.0))
    # The segment is relative to the agent too: a vertical face at x = 1.9.
    assert nearest.segment.p1[0] == pytest.approx(1.9)
    assert nearest.segment.p2[0] == pytest.approx(1.9)

File: libsimulator/test/TestAgentView.cpp
// SPDX-License-Identifier: LGPL-3.0-or-later
#include "AgentView.hpp"
#include "EnvironmentQuery.hpp"
#include "GenericAgent.hpp"
#include "Geometry/Geometry3D.hpp"
#include "GeometryBuilder.hpp"
#include "MeshFixtures.hpp"
#include "NeighborhoodSearch.hpp"
#include "OperationalModels/CollisionFreeSpeedModel/CollisionFreeSpeedModel.hpp"
#include "TestCommon.hpp"

#include <gtest/gtest.h>

#include <algorithm>
#include <limits>
#include <memory>
#include <vector>

namespace
{
using State = CollisionFreeSpeedModel::State;

GenericAgent MakeAgent(const Geometry3D& geo, Point pos, double radius = 0.2, double z = 0.0)
{
    State s{};
    s.radius = radius;
    return GenericAgent(
        GenericAgent::ID::Invalid,
        jps::UniqueID<Journey>::Invalid,
        jps::UniqueID<BaseStage>::Invalid,
        *geo.get_location(pos.x, pos.y, z),
        std::move(s));
}

std::unique_ptr<Geometry3D> OpenGeometry()
{
    GeometryBuilder b{};
    b.AddAccessibleArea({{-100, -100}, {100, -100}, {100, 100}, {-100, 100}});
    return std::make_unique<Geometry3D>(b.Build().Polygon());
}

// Geometry with a thin wall at x≈1 that blocks line-of-sight across it.
std::unique_ptr<Geometry3D> WalledGeometry()
{
    GeometryBuilder b{};
    b.AddAccessibleArea({{-100, -100}, {100, -100}, {100, 100}, {-100, 100}});
    b.ExcludeFromAccessibleArea({{0.9, -50}, {1.1, -50}, {1.1, 50}, {0.9, 50}});
    return std::make_unique<Geometry3D>(b.Build().Polygon());
}

struct Environment {
    AgentContainer<GenericAgent> agents{};
    NeighborhoodSearch<GenericAgent> neighborhood_search{5.0};
    std::vector<std::pair<Point, double>> requested{};

    void add_agent(Point pos, double radius = 0.2) { requested.emplace_back(pos, radius); }

    // Agents are asked for before the geometry exists, so they are made here -- only the
    // geometry can put them onto the surface, the same way Simulation::AddAgent does.
    EnvironmentQuery query(const Geometry3D& geo)
    {
        for(const auto& [pos, radius] : requested) {
            agents.push_back(MakeAgent(geo, pos, radius));
        }
        requested.clear();
        neighborhood_search.Update(agents);
        return {geo, neighborhood_search};
    }

    // The first agent added is the one every test queries from.
    AgentView FirstAgentView(const EnvironmentQuery& q) const { return {q, agents[0]}; }

    const Location& FirstAgentLocation() const { return agents[0].location; }
};
} // namespace

TEST(AgentView, OtherAgentsInRangeExcludesSelf)
{
    Environment env{};
    env.add_agent({0, 0});
    const auto geo = OpenGeometry();
    const auto q = env.query(*geo);

    const auto result = env.FirstAgentView(q).OtherAgentsInRange(100.0);
    EXPECT_TRUE(result.empty());
}

TEST(AgentView, OtherAgentsInRangeNoFilterReturnsAllInRadius)
{
    Environment env{};
    env.add_agent({0, 0}); // querying agent
    env.add_agent({1, 0});
    env.add_agent({0, 1});
    env.add_agent({-1, 0});
    const auto geo = OpenGeometry();
    const auto q = env.query(*geo);

    const auto result = env.FirstAgentView(q).OtherAgentsInRange(5.0);
    EXPECT_EQ(result.size(), 3u);
}

TEST(AgentView, OtherAgentsInRangeCustomFilterRejectsAll)
{
    Environment env{};
    env.add_agent({0, 0});
    env.add_agent({1, 0});
    env.add_agent({0, 1});
    const auto geo = OpenGeometry();
    const auto q = env.query(*geo);

    const auto result =
        env.FirstAgentView(q).OtherAgentsInRange(5.0, [](const NeighborView&) { return false; });
    EXPECT_TRUE(result.empty());
}

TEST(AgentView, OtherAgentsInRangeCustomFilterSelectsSubset)
{
    Environment env{};
    env.add_agent({0, 0}); // querying agent
    env.add_agent({1, 0}); // positive x — kept
    env.add_agent({0, 1}); // positive y — kept
    env.add_agent({-1, 0}); // negative x — filtered out
    const auto geo = OpenGeometry();
    const auto q = env.query(*geo);

    const auto result = env.FirstAgentView(q).OtherAgentsInRange(
        5.0, [](const NeighborView& n) { return n.RelativePosition.x >= 0.0; });

    ASSERT_EQ(result.size(), 2u);
    for(const auto& neighbor : result) {
        EXPECT_GE(neighbor.RelativePosition.x, 0.0);
    }
}

TEST(AgentView, NoGeometryBetweenFiltersOccludedAgents)
{
    Environment env{};
    env.add_agent({0, 0}); // querying agent
    env.add_agent({2, 0}); // behind wall — occluded
    env.add_agent({0, 1}); // same side as querying agent — visible
    const auto geo = WalledGeometry();
    const auto q = env.query(*geo);

    const auto view = env.FirstAgentView(q);
    const auto result = view.OtherAgentsInRange(
        5.0, [&](const NeighborView& n) { return view.NoGeometryBetween(n); });

    ASSERT_EQ(result.size(), 1u);
    EXPECT_EQ(result[0].RelativePosition, Point(0, 1));
}

TEST(AgentView, OtherAgentsInRangeCustomFilterReceivesNoSelf)
{
    // Verify the filter is never called with the querying agent itself.
    Environment env{};
    env.add_agent({0, 0});
    env.add_agent({1, 0});
    const auto geo = OpenGeometry();
    const auto q = env.query(*geo);

    int calls = 0;
    env.FirstAgentView(q).OtherAgentsInRange(5.0, [&](const NeighborView&) {
        ++calls;
        return true;
    });
    EXPECT_EQ(calls, 1);
}

TEST(AgentView, OtherAgentsInRangeOutOfRadiusNotReturned)
{
    Environment env{};
    env.add_agent({0, 0});
    env.add_agent({50, 0}); // far away
    const auto geo = OpenGeometry();
    const auto q = env.query(*geo);

    const auto result =
        env.FirstAgentView(q).OtherAgentsInRange(1.0, [](const NeighborView&) { return true; });
    EXPECT_TRUE(result.empty());
}

TEST(AgentView, AgentsOnTheSamePositionSeeEachOther)
{
    Environment env{};
    env.add_agent({3, 4});
    env.add_agent({3, 4});
    const auto geo = OpenGeometry();
    const auto q = env.query(*geo);

    const auto result = env.FirstAgentView(q).OtherAgentsInRange(1.0);

    ASSERT_EQ(result.size(), 1u);
    EXPECT_EQ(result[0].state, &env.agents[1].state);
}

// Every other test here queries from {0,0} and would not notice a centre that
// silently defaults to the origin.
TEST(AgentView, OtherAgentsInRangeCentresOnTheQueryingAgentNotTheOrigin)
{
    Environment env{};
    env.add_agent({50, 30}); // querying agent, deliberately away from the origin
    env.add_agent({51, 30}); // its actual neighbor
    env.add_agent({0.5, 0}); // decoy next to the origin
    const auto geo = OpenGeometry();
    const auto q = env.query(*geo);

    const auto result = env.FirstAgentView(q).OtherAgentsInRange(2.0);

    ASSERT_EQ(result.size(), 1u);
    EXPECT_EQ(result[0].RelativePosition, Point(1, 0));
}

TEST(AgentView, WallsInRangeAreRelativeToTheAgent)
{
    Environment env{};
    const Point pos{-1.0, 2.0};
    env.add_agent(pos);
    const auto geo = WalledGeometry();
    const auto q = env.query(*geo);

    std::vector<LineSegment> expected{};
    for(const auto& segment : q.LineSegmentsInRange(env.FirstAgentLocation(), 3.0)) {
        expected.push_back({segment.p1 - pos, segment.p2 - pos});
    }
    ASSERT_FALSE(expected.empty());

    const auto view = env.FirstAgentView(q);
    std::vector<LineSegment> actual{};
    double closest = std::numeric_limits<double>::max();
    for(const auto& wall : view.WallsInRange(3.0)) {
        actual.push_back(wall.segment);
        closest = std::min(closest, wall.distance);
    }

    EXPECT_EQ(actual, expected);

    // The near face of the wall block sits at x = 0.9, so measured from the agent it is
    // 1.9 away. Anything computed against the origin instead would report 0.9.
    EXPECT_DOUBLE_EQ(closest, 1.9);
}

TEST(AgentView, WallViewProjectsOntoTheWallAndPointsBackAtTheAgent)
{
    Environment env{};
    const Point pos{-1.0, 2.0};
    env.add_agent(pos);
    const auto geo = WalledGeometry();
    const auto q = env.query(*geo);

    const auto view = env.FirstAgentView(q);
    auto walls = view.WallsInRange(3.0);
    ASSERT_FALSE(walls.empty());
    // The near face of the wall block spans the agent's y, so the agent faces it head on.
    const WallView nearest =
        *std::ranges::min_element(walls, {}, [](const WallView& w) { return w.distance; });

    EXPECT_DOUBLE_EQ(nearest.closest_point.x, 1.9);
    EXPECT_NEAR(nearest.closest_point.y, 0.0, 1e-12);
    EXPECT_DOUBLE_EQ(nearest.distance, 1.9);
    // The wall is to the agent's right, so a normal pointing back at the agent faces -x.
    EXPECT_DOUBLE_EQ(nearest.normal.x, -1.0);
    EXPECT_NEAR(nearest.normal.y, 0.0, 1e-12);
}

TEST(AgentView, AgentsOnAnotherStoreyAreNotNeighbours)
{
    // Two floors sharing a footprint, one agent on each, standing at the same (x, y) three
    // metres apart in height. The neighbourhood grid searches by (x, y) and offers them to
    // each other; one is standing well above the other's head and cannot touch them.
    Geometry3D geo{fixtures::stacked_floors()};

    AgentContainer<GenericAgent> agents{};
    agents.push_back(MakeAgent(geo, {5.0, 5.0}));
    agents.push_back(MakeAgent(geo, {5.0, 5.0}, 0.2, 3.0));

    NeighborhoodSearch<GenericAgent> search{5.0};
    search.Update(agents);
    const EnvironmentQuery query{geo, search};

    EXPECT_TRUE(AgentView(query, agents[0]).OtherAgentsInRange(10.0).empty());
    EXPECT_TRUE(AgentView(query, agents[1]).OtherAgentsInRange(10.0).empty());
}

TEST(AgentView, AgentsOnTheSameStoreyStillAre)
{
    // The same mesh, both agents on the lower floor: the filter must not swallow those.
    Geometry3D geo{fixtures::stacked_floors()};

    AgentContainer<GenericAgent> agents{};
    agents.push_back(MakeAgent(geo, {5.0, 5.0}));
    agents.push_back(MakeAgent(geo, {6.0, 5.0}));

    NeighborhoodSearch<GenericAgent> search{5.0};
    search.Update(agents);
    const EnvironmentQuery query{geo, search};

    EXPECT_EQ(AgentView(query, agents[0]).OtherAgentsInRange(10.0).size(), 1u);
}

TEST(AgentView, ANeighbourCloseEnoughToTouchCanStillBeOnAnotherStorey)
{
    // A mezzanine 1.5 m up, so the height band keeps both as candidates and what has to tell
    // them apart is which sheet each stands on. Nothing is between them in plan either: no
    // wall on either floor, and no seam joining the two.
    Geometry3D geo{fixtures::stacked_floors(1.5)};

    AgentContainer<GenericAgent> agents{};
    agents.push_back(MakeAgent(geo, {2.0, 5.0}));
    agents.push_back(MakeAgent(geo, {4.0, 5.0}));
    agents.push_back(MakeAgent(geo, {6.0, 5.0}, 0.2, 1.5));

    NeighborhoodSearch<GenericAgent> search{5.0};
    search.Update(agents);
    const EnvironmentQuery query{geo, search};

    const AgentView view{query, agents[0]};
    const auto seen = view.OtherAgentsInRange(
        10.0, [&](const NeighborView& n) { return view.NoGeometryBetween(n); });

    ASSERT_EQ(seen.size(), 1u);
    EXPECT_EQ(seen[0].RelativePosition, Point(2.0, 0.0));
}
